snakesee/utils.py

```
@@ -199,19 +199,115 @@ def safe_file_size(path: Path) -> int:
-        directory: Directory to scan.
+class _ScanCache:
+    """Cache for directory scan results to avoid re-stat-ing thousands of files.
 
-    Returns:
-        List of _MetadataFileInfo for all files found.
+    Tracks directory mtimes to detect when new files are added. Since metadata
+    files are append-only (Snakemake adds them, never modifies or deletes),
+    we only need to rescan directories whose mtime has changed.
     """
+
+    __slots__ = ("_files", "_dir_mtimes", "_lock")
```

Ruff (0.15.7)

[warning] `_ScanCache.__slots__` is not sorted

Apply a natural sort to `_ScanCache.__slots__`

(RUF023)